### ISSUE TITLE
Update doxygen package for Ubuntu 18.04

### ISSUE THREAD

**Issue by ls6777** [doxygen/doxygen#8054]
**Describe the bug**
I've tried for hours to compile, install, use sources, use binaries, to get doxygen 1.8.20 and 1.8.17 to work on Ubuntu 18.04 distro.  I have failed.  The binaries for 1.8.17 and 1.8.20 do not work on Ubuntu 18.04 and I have been unsuccessful in building it for my distro.  It would be very helpful if the official package install for doxygen (apt-get install doxygen) was the latest version, or just newer than 1.8.13.

**Expected behavior**
"apt-get install doxygen" install version 1.8.20

**To Reproduce**
"apt-get install doxygen"

**Version**
Docker container utilizing Ubuntu 18.04 distro

**Comment by albert-github** [comment]
- The "apt-get install doxygen" is outside the control of doxygen and should directly be directed to the Ubuntu package maintainers.

- Regarding the "to compile, install, use sources, use binaries, to get doxygen 1.8.20", what did you exactly try
  - `git clone` or `tar xf source` ...
  - `cmake` ....
  - `make`  ...

  what were the error messages you got?

**Comment by ls6777** [comment]
Thanks for the info @albert-github about the package maintainers.  I checked out this site for that:
https://tracker.debian.org/pkg/doxygen

Seems like they have been having issues since 1.8.13 to get stable build.

![image](https://user-images.githubusercontent.com/26239627/94157235-d9dbbd80-fe35-11ea-897a-31c0b208d889.png)

I will update shortly about 1.8.20 build process.

**Comment by ls6777** [comment]
These are the step I follow to install doxygen 1.8.20:
` apt-get install graphviz`
`apt-get bison`
`apt-get install flex`
`wget http://ftp.gnu.org/pub/gnu/libiconv/libiconv-1.16.tar.gz`
`tar -xvzf libiconv-1.16.tar.gz`
`cd libiconv-1.16`
`./configure --prefix=/usr/local/libiconv`
`make`
`make install`
`wget http://doxygen.nl/files/doxygen-1.8.20.src.tar.gz`
`gunzip doxygen-1.8.20.src.tar.gz`
`tar -xf doxygen-1.8.20.src.tar`
`cd doxygen-1.8.20`
`mkdir build`
`cd build`
`cmake -G "Unix Makefiles" -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/usr -Duse_libclang=ON ..`

At this point I get the following output:
```
CMake Error at CMakeLists.txt:51 (find_package):
  Could not find a package configuration file provided by "LLVM" with any of
  the following names:

    LLVMConfig.cmake
    llvm-config.cmake

  Add the installation prefix of "LLVM" to CMAKE_PREFIX_PATH or set
  "LLVM_DIR" to a directory containing one of the above files.  If "LLVM"
  provides a separate development package or SDK, be sure it has been
  installed.


-- Configuring incomplete, errors occurred!
See also "/home/doxygen-1.8.20/build/CMakeFiles/CMakeOutput.log".
```
Now I get LLVM package:
`apt-get install llvm`
`cmake -G "Unix Makefiles" -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/usr -Duse_libclang=ON ..`

At this point I get the following output:
```
CMake Error at CMakeLists.txt:52 (find_package):
  Could not find a package configuration file provided by "Clang" with any of
  the following names:

    ClangConfig.cmake
    clang-config.cmake

  Add the installation prefix of "Clang" to CMAKE_PREFIX_PATH or set
  "Clang_DIR" to a directory containing one of the above files.  If "Clang"
  provides a separate development package or SDK, be sure it has been
  installed.


-- Configuring incomplete, errors occurred!
```

Try getting clang package:
`apt-get install clang`

Get the same error
```
CMake Error at CMakeLists.txt:52 (find_package):
  Could not find a package configuration file provided by "Clang" with any of
  the following names:

    ClangConfig.cmake
    clang-config.cmake

  Add the installation prefix of "Clang" to CMAKE_PREFIX_PATH or set
  "Clang_DIR" to a directory containing one of the above files.  If "Clang"
  provides a separate development package or SDK, be sure it has been
  installed.


-- Configuring incomplete, errors occurred!
See also "/home/doxygen-1.8.20/build/CMakeFiles/CMakeOutput.log".

Now what?
```

**Comment by albert-github** [comment]
LLVM and Clang giving sometimes some problems (by head I think in the current master some fixes have been made in respect to the build with clang). So I would, just for the first tests, just skip them. Just set `-Duse_libclang=OFF` and see how far you get at that point. After we have succeeded in getting a version without Clang we can have a look at the problems with Clang.

**Comment by ls6777** [comment]
I can get it working fine without Clang.  No problem there.  
However, my configuration uses Clang, so that's where I have been getting stuck.

**Comment by albert-github** [comment]
OK I understand, this is already a good start. As written there are some fixes already in the master version, lets see whether this fixes the CLang part as well, so instead of:
```
wget http://doxygen.nl/files/doxygen-1.8.20.src.tar.gz
gunzip doxygen-1.8.20.src.tar.gz
tar -xf doxygen-1.8.20.src.tar
cd doxygen-1.8.20
```
just do:
```
git clone https://github.com/doxygen/doxygen.git
cd doxygen 
```
and create a version with the `-Duse_libclang=ON`

**Comment by ls6777** [comment]
`git clone https://github.com/doxygen/doxygen.git`
`cd doxygen`
`mkdir build`
`cd build`
`cmake -G "Unix Makefiles" -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/usr -Duse_libclang=ON ..`

Same error:
```
-- The C compiler identification is GNU 7.5.0
-- The CXX compiler identification is GNU 7.5.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at CMakeLists.txt:52 (find_package):
  Could not find a package configuration file provided by "Clang" with any of
  the following names:

    ClangConfig.cmake
    clang-config.cmake

  Add the installation prefix of "Clang" to CMAKE_PREFIX_PATH or set
  "Clang_DIR" to a directory containing one of the above files.  If "Clang"
  provides a separate development package or SDK, be sure it has been
  installed.


-- Configuring incomplete, errors occurred!
See also "/home/doxygen/build/CMakeFiles/CMakeOutput.log".
```

**Comment by albert-github** [comment]
I'm not an expert for CLang / LLVM but 
- is either of the files ClangConfig.cmake or clang-config.cmake or something similar preset on your system? (on Cygwin I found lib/cmake/clang/ClangConfig.cmake and some more system specific files under usr/share/cmake-3.14.5/Modules, I know Cygwin is completely different but it might give a clue). When present setting build_windows_nmake_v2017_clang might help.
- does Ubuntu  have separate development package or SDK, if so installing might help

Some verion information:
- Which version of LLVM / CLang did you install?
- which version of CMake do you use?

**Comment by ls6777** [comment]
```
root@45a12a22d1ad:/home/doxygen/build# clang --version
clang version 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```
```
root@45a12a22d1ad:/home/doxygen/build# llvm-config --version
6.0.0
```
```
root@45a12a22d1ad:/home/doxygen/build# find /usr/ -name ClangConfig.cmake
/usr/share/llvm-6.0/cmake/ClangConfig.cmake
```
Tried the following:
`cmake -G "Unix Makefiles" -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/usr -DClang_DIR=/usr/share/llvm-6.0/cmake -Duse_libclang=ON ..`

Got the following output:
```
CMake Error at /usr/share/llvm-6.0/cmake/ClangConfig.cmake:18 (include):
  include could not find load file:

    /usr/lib/cmake/clang/ClangTargets.cmake
Call Stack (most recent call first):
  CMakeLists.txt:52 (find_package)


-- Found PythonInterp: /usr/bin/python (found version "2.7.17")
-- Found FLEX: /usr/bin/flex (found version "2.6.4")
-- Found BISON: /usr/bin/bison (found version "3.0.4")
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Looking for iconv_open
-- Looking for iconv_open - found
-- Performing Test ICONV_COMPILES
-- Performing Test ICONV_COMPILES - Success
-- Found Iconv: In glibc
-- One (and only one) of the ICONV_ACCEPTS_... tests must pass
-- Performing Test ICONV_ACCEPTS_NONCONST_INPUT
-- Performing Test ICONV_ACCEPTS_NONCONST_INPUT - Success
-- Performing Test ICONV_ACCEPTS_CONST_INPUT
-- Performing Test ICONV_ACCEPTS_CONST_INPUT - Failed
-- The javacc executable not found, using existing files
CMake Error at /usr/share/llvm-6.0/cmake/ClangConfig.cmake:18 (include):
  include could not find load file:

    /usr/lib/cmake/clang/ClangTargets.cmake
Call Stack (most recent call first):
  src/CMakeLists.txt:328 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/doxygen/build/CMakeFiles/CMakeOutput.log".
See also "/home/doxygen/build/CMakeFiles/CMakeError.log".
```

**Comment by albert-github** [comment]
Hmm, does this ClangTargets.cmake exist on your system? Maybe you also need to set the LLVM_DIR (but that would not be very logical).

@doxygen is version 6 OK?

**Comment by ls6777** [comment]
Tried the following:
`apt-get install clang-format clang-tidy clang-tools clang libc++-dev libc++1 libc++abi-dev libc++abi1 libclang-dev libclang1 libomp-dev libomp5 lld lldb llvm-dev llvm-runtime llvm`

Then:
`cmake -G "Unix Makefiles" -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/usr -Duse_libclang=ON ..`

Output:
```
CMake Error at /usr/share/llvm-6.0/cmake/ClangConfig.cmake:18 (include):
  include could not find load file:

    /usr/lib/cmake/clang/ClangTargets.cmake
Call Stack (most recent call first):
  CMakeLists.txt:52 (find_package)


-- The javacc executable not found, using existing files
CMake Error at /usr/share/llvm-6.0/cmake/ClangConfig.cmake:18 (include):
  include could not find load file:

    /usr/lib/cmake/clang/ClangTargets.cmake
Call Stack (most recent call first):
  src/CMakeLists.txt:328 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/doxygen/build/CMakeFiles/CMakeOutput.log".
See also "/home/doxygen/build/CMakeFiles/CMakeError.log".
```

Check for ClangTargets.cmake:
`find /usr/ -name ClangTargets.cmake`

Received:
`/usr/share/llvm-6.0/cmake/ClangTargets.cmake`

It's in the same directory as ClangConfig.cmake, so I don't know why it can'd find it.

Do I need to modify the cmake files?

**Comment by albert-github** [comment]
To the best of my knowledge you don't have to update cmake files or PATH.

- Nice set of `apt-get`, did the file `ClangTargets.cmake` enter now or was it already present?
- Why didn't you use `Clang_DIR` anymore?

**Comment by ls6777** [comment]
`ClangTargets.cmake` was already present

I tried to build with both `Clang_DIR` and without.  Getting the same results.

Found this issue with google search: [https://github.com/woboq/woboq_codebrowser/issues/74](https://github.com/woboq/woboq_codebrowser/issues/74)

Going to try clang-7 and report back.

**Comment by albert-github** [comment]
Nice find /search.

I don't know which versions of LLVM / Clang are available for your system, but the current version is already a bit newer, so when a newer version is available I would go for that newer version (they are working on version 11 and version 10.0.1 is the latest released version).

**Comment by ls6777** [comment]
Success!!

`apt install llvm-10 clang-10 libclang-10-dev`

```
root@89cf692b5e93:/home/libiconv-1.16/doxygen-1.8.20/build# cmake -G "Unix Makefiles" -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/usr -Duse_libclang=ON ..
-- Found PythonInterp: /usr/bin/python (found version "2.7.17")
-- Found FLEX: /usr/bin/flex (found version "2.6.4")
-- Found BISON: /usr/bin/bison (found version "3.0.4")
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Looking for iconv_open
-- Looking for iconv_open - found
-- Performing Test ICONV_COMPILES
-- Performing Test ICONV_COMPILES - Success
-- Found Iconv: In glibc
-- One (and only one) of the ICONV_ACCEPTS_... tests must pass
-- Performing Test ICONV_ACCEPTS_NONCONST_INPUT
-- Performing Test ICONV_ACCEPTS_NONCONST_INPUT - Success
-- Performing Test ICONV_ACCEPTS_CONST_INPUT
-- Performing Test ICONV_ACCEPTS_CONST_INPUT - Failed
-- The javacc executable not found, using existing files
-- Configuring done
-- Generating done
-- Build files have been written to: /home/libiconv-1.16/doxygen-1.8.20/build
```
Running `make` was successful
Running `make install` was successful

I will probably put a write up question and answer in stack overflow in case someone else runs into this issue.

**Comment by ls6777** [comment]
Not sure why `apt-get install llvm` and `apt-get install clang` installs version 6... you would think it would install the latest version.

**Comment by savoury1** [comment]
Although this issue is closed, in case anyone else searches for the CMake error encountered by OP (as I did this week) and lands here then some extra information might be helpful. As mentioned in the issue OP linked at https://github.com/doxygen/doxygen/issues/8054#issuecomment-698480600 this is a packaging bug with earlier revisions of Debian/Ubuntu LLVM 6.0, see https://github.com/woboq/woboq_codebrowser/issues/74#issuecomment-493736735.

Having encountered the same CMake error trying to build doxygen 1.8.17 for older Ubuntu series (ie. Xenial & Bionic) in the past few days led me to this issue here. Although older package revisions of LLVM 6.0 for Debian/Ubuntu are broken for building with CMake (not finding Clang) any llvm-toolchain-6.0 packages with version 1:6.0.1~+rc1-1~exp2 (7 May 2018) and higher fix the CMake breakage relative Clang. Thus, a backport of an LLVM 6.0 package with at least that Debian package revision or higher will work fine for building doxygen (and whatever else that might require Clang 6.0) with CMake.

Relative `apt-get install llvm` not installing the latest LLVM version, this is due the way that "defaults" work on Linux-based systems. There is often a secondary package, in this case llvm-defaults for Debian/Ubuntu systems, that defines the desired default version of the software (whether LLVM, GCC, etc). To make the local system use a different default version either requires setting up alternatives oneself (see "man update-alternatives") or building and installing an updated defaults package.

Anyone wanting an LLVM 6.0 for either Ubuntu Xenial or Bionic that works with CMake can see my [PPA](https://launchpad.net/~savoury1/+archive/ubuntu/llvm-toolchain) where I have today backported LLVM 6.0.1 from Focal to Xenial & Bionic. This backported version works fine to build doxygen 1.8.17 successfully (as I have now been able to do). Building doxygen with LLVM 6 avoids bringing in dependencies on LLVM 10 packages, which older series (ie. Xenial) don't have available by default (although Bionic does now have LLVM 10 available in the official bionic-updates repo). The linked PPA also has backports of LLVM 8.0.1 and 9.0.1 for Xenial & Bionic systems, as well as 10.0.0 for Xenial systems.
